libqtile/group.py

```diff
@@ -83,23 +85,54 @@ def current_window(self):
 
     @current_window.setter
     def current_window(self, win):
-        try:
+        old = self._order_clients()
+        if win in self.focus_history:
             self.focus_history.remove(win)
-        except ValueError:
-            # win has never received focus before
-            pass
+        self.windows.remove(win)
+
         self.focus_history.append(win)
+        self.windows.append(win)
+        if old != self._order_clients():
+            self.assure_correct_layer(win)
 
-    def _remove_from_focus_history(self, win):
-        try:
-            index = self.focus_history.index(win)
-        except ValueError:
-            # win has never received focus
-            return False
+    def _order_clients(self):
+        """
+        Orders our clients by their z-value, leaving the order as is in case of equal z.
+        """
+        def sort_by(client):
+            if client.floating and not client.fullscreen:
+                return client.z + window.FLOAT_LAYER
+            return client.z
+        return sorted(filter(lambda w: not w.minimized, self.windows), key=sort_by)
+
+    def assure_correct_layer(self, win, **kwargs):
+        """
+        Make sure that the window is in its designated z-position by moving other windows around as necessary.
+        We assume that win is already at the correct position in our list of windows. Nobody else should be
+        calling xcbq.configure with a StackMode (except Popups).
+        """
+        if win.minimized:
+            return
+
+        stacked = ['max']  # z-order in these layouts does matter
```

Comment:
This is quite the question. I started by doing the layering in `Group.layout_all` following a suggestion by ramnes . To avoid some flickering I switched to reacting to changes of the `current_window` later. But there are points to be made to move the z-management up and/or down the chain of responsibility: The layout should handle its windows, but then you can't force a normal window to stay on top of a floating window, only the group can do that. The group really can't access the bar or other windows outside its realm of responsibility, so maybe the z-layering should be done on a screen-level.

I'll have to think about how to properly "layer" this entire concept of z-values, ideally everybody can take care of its scope itself, but a group should be able to overrule a layout and a screen to overrule a group, if needed. I am fairly sure there is a lot of room for improvement, but it would also mean a lot more changes to the current code.

Reply:
After having learned quite a bit about layouts, I have thought about approaching this a little differently to involve the layouts more:

If the layouts got their own z-level (keeping floating above everything else might be sufficient for now), and a way to track the z-levels of clients, they could pass a client's current z-level to `Window.place()`, which every layout calls when asked to `configure` a client. The group would then realize the suggestions of its layouts.
This would remove the ability to set custom z-levels, be it by the user or through `update_state` - those could maybe be addressed by another z-related property on windows (e.g. `z_offset_user` or `z_offset_wm`?).

The group would still be the executive force, but the layouts could give an opinion on where windows should go. The group wouldn't really have to track z-levels itself anymore, just inquire the suggested order from its layouts.

I'll try to implement that approach in the next few days to see where it leads.

Reply:
OK I've read these comments over and over trying to organise my thoughts :P 

I think the screen shouldn't have anything to do with it. Screens are somewhat above groups and and considering a group _is_ is the collection of windows, it does make sense for Z control to be handled (mostly) at the group level. Consider if you were to move a group to a different screen - you would expect the relative Z positions of its windows to be unchanged.

So that takes us to groups and layouts. I agree with your idea: layouts can have Z control over the tiled windows that belong to that layout. The layout is, in a sense, a tool used by the group to handle a subset of the groups windows. The remainder are floating windows handled directly by the group. I think that ideally a layout would have absolute control over its windows' Z position, much like they are over their windows' X and Y positions. If a user modifies a tiled window's X or Y (by dragging, for example), that window is yanked out of the layout and becomes floating, so the same thing should happen when the Z value is modified by a user.

Does that make sense?

Reply:
Thanks for your input. This is a bit of a complex topic. :) 

> it does make sense for Z control to be handled (mostly) at the group level

You are leaving a bit of room here: If not for `Screens`, then what should manage the layering for windows outside of groups in your opinion? Should it be managed at all?

> The remainder are floating windows handled directly by the group. [...]
> If a user modifies a tiled window's X or Y (by dragging, for example), that window is yanked out of the layout and becomes floating, so the same thing should happen when the Z value is modified by a user.

Well, there is a floating layout managing the groups floating windows. Continuing your thoughts, that would mean that the floating layout mustn't manage z-layers (we have to assume z to be user-modified), right?
For the user, there is not much of a way to fine-control z (as opposed to x and y) - even if additional functions were provided. So maybe some distinction between (x, y) and (z) is advisable.

Reply:
I added a number of changes, some of which reverting previous changes. Therefore I also cleaned up the commit history again, followed by a forced push. I hope that's ok, since it "detaches" a lot of comments in here.

The relevant layouts (`Max` and `Floating`) now suggest a z-level for their clients. The group is then asked to enforce that in `layout_all`. The code has gotten simpler because of it, so I have a strong feeling that this is the right direction.

Reply:
Also, the hard reference to `Max` has been removed. I'll mark this conversation resolved in a few days, unless something new comes up.